docs/source/profiling.rst

```diff
@@ -0,0 +1,498 @@
+``performance.sqlite`` in the run directory, next to the main log file. This
+file is a database containing several tables with performance information about
+the run.
+
+This information can be analysed in three ways: 1) through the ``muscle3
+profile`` command, 2) via the :class:`libmuscle.ProfileDatabase` class in
+Python, and 3) directly via ``sqlite3`` from the shell or from Python. All three
+are described here.
+
+
+Plotting statistics from the command line
+-----------------------------------------
+
+The most simplest way of examining performance data gathered by MUSCLE3 is
+through the ``muscle3 profile`` command from the shell. If you have done a run,
+then you should have a run directory containing a ``performance.sqlite`` file.
+If you have MUSCLE3 available in your environment (only the Python installation
+is needed) then you have the ``muscle3 profile`` command available to show
+per-instance and per-core statistics as well as a timeline of events.
+
+Per-instance time spent
+```````````````````````
+
+.. image:: plot_performance_instances.png
+   :width: 80%
+   :align: center
+
+.. code-block:: bash
+
+    muscle3 profile --instances /path/to/performance.sqlite
+
+With ``--instances``, the plot will show for each instance how much time it
+spent in total on computing, communicating and waiting. This plot gives an idea
+of where most of the computing is done, and which components you need to
+optimise to get an answer sooner.
+
+In many models, you will find that there's one component that takes up most of
+the compute time, and others that spend most of their time waiting and then do a
+little bit of computing. This is not necessarily wasteful, it may simply be that
+most of the work is in that one component. Of course it does mean that if you
+want to get your result faster, you should focus your efforts on the component
+that takes the most time.
+
+This works fine for relatively simple models, but there are some caveats to take
+into account. First, model components may use different amounts of resources. A
+component that runs 10% of the time, but uses 100 cores, uses a lot more energy
+than one that runs 90% of the time on a single core. On the other hand, if you
+are interested in getting an answer more quickly, then you should optimise the
+single-core model first.
+
+If you are running on HPC, then you most likely have a fixed budget of
+core hours or system billing units (SBUs), and reducing SBU usage may be of
+interest. In the above scenario, both models are actually candidates for
+optimisation in this case. If you can speed up the parallel model so that it can
+run as fast on fewer cores then that potentially saves many SBUs. But speeding
+up the single-thread model also helps, because even if the cores that the
+parallel model runs on aren't used, they're still allocated and you're still
+paying for them.
+
+If all your models spend most of their time waiting, then your models do
```

Comment:
Hmm, good point. The core issue here is that the total time spent computing is low compared to the time spent sending and receiving, because that means there's a lot of communication overhead. I'll rephrase.